### DISCUSSION TITLE
beginner question on calling external URLs

### DISCUSSION THREAD

**Discussion by JLT032** [rwf2/Rocket#1983] [Questions]
Hey,

I'm making a simple app with rocket which takes an URL/value/value/value
where the value should then query an external url URL2/api/call/value/value/value and return the result

this result is in JSON which i then hope to temporarily store and reformat for easy viewing

now, the main problem is i don't find a way to make the query to URL2 with what i learned about rocket, once that is done i think i'll be able to continue on my own

**Comment by somehowchris** [comment]
@commandline-be If that URL2 is on an external socker, use something like [reqwest](https://docs.rs/reqwest/0.11.6/reqwest/), if it is on the same socker (i.e. same domain, different path) use the [`uri!` macro](https://rocket.rs/v0.5-rc/guide/responses/#typed-uris)

**Comment by JLT032** [comment]
Thanks @somehowchris 

My approach would be both i guess.

1. obtain data from external sources
2. present this data using rocket.rs

How think about this is 

1. query URL to check if data needs to be download
2. if 1==true download the data
3. convert the data into JSON and/or absorb into SQL database
4. use rocket to present the data from JSON/SQL

it is clear i'm moving cautiously here, this is unknown territory but it is fun learning

**Reply by somehowchris** [reply]
@commandline-be 

First of all, caching is hard af to get right and only should be done if considered really carefully especially if you don't know what data will be served (nature, structure, integrety, quality,.. of the data)

As mentioned above use [`reqwest`](https://docs.rs/reqwest/0.11.6/reqwest/) to make http requests to/against a socket external api.

Doing a micro-service-ish architecture/request proxy you should only cache hot data if you store any data. Again doing caching on unpredicable data is hard. 

If this is a research or personal project your solution can be fun and enough. If you are considering to go prod or use it as an acutal example go for something like redis, etcd or just plain in memory states. They are quick, usualy key-value which means you can hash something as a key and save the json as the value and scale much better then parsing and quering acual sql databases
